hibernate-core/src/main/java/org/hibernate/internal/util/collections/StandardStack.java

```diff
@@ -6,139 +6,159 @@
 import java.util.function.BiFunction;
 import java.util.function.Consumer;
 import java.util.function.Function;
 
 /**
  * A general-purpose stack impl supporting null values.
+ * <p>
+ * Most of the implementation was inspired by {@link ArrayDeque} methods
+ * and follows the same logic.
  *
  * @param <T> The type of things stored in the stack
  *
  * @author Steve Ebersole
  * @author Sanne Grinovero
+ * @author Marco Belladelli
  */
 public final class StandardStack<T> implements Stack<T> {
+	private T[] elements;
+	private int head = 0;
+	private int tail = 0;
 
-	private ArrayDeque<T> internalStack;
-	private static final Object NULL_TOKEN = new Object();
-
-	public StandardStack() {
+	@SuppressWarnings("unchecked")
+	public StandardStack(Class<T> type) {
+		elements = (T[]) Array.newInstance( type, 8 );
 	}
 
-	public StandardStack(T initial) {
-		stackInstanceExpected().addFirst( initial );
+	public StandardStack(Class<T> type, T initial) {
+		this( type );
+		push( initial );
 	}
 
 	@Override
-	public void push(T newCurrent) {
-		T toStore = newCurrent;
-		if ( newCurrent == null ) {
-			toStore = (T) NULL_TOKEN;
-		}
-		stackInstanceExpected().addFirst( toStore );
-	}
-
-	private Deque<T> stackInstanceExpected() {
-		if ( internalStack == null ) {
-			//"7" picked to use 8, but skipping the odd initialCapacity method
-			internalStack = new ArrayDeque<>( 7 );
+	public void push(T e) {
+		head = dec( head );
+		elements[head] = e;
+		if ( head == tail ) {
+			grow();
 		}
-		return internalStack;
 	}
 
 	@Override
 	public T pop() {
-		return convert( stackInstanceExpected().removeFirst() );
-	}
-
-	private T convert(final Object internalStoredObject) {
-		if ( internalStoredObject == NULL_TOKEN ) {
-			return null;
+		if ( isEmpty() ) {
+			throw new NoSuchElementException();
 		}
-		return (T) internalStoredObject;
+		T e = elements[head];
+		elements[head] = null;
+		head = inc( head );
+		return e;
 	}
 
 	@Override
 	public T getCurrent() {
-		if ( internalStack == null ) {
-			return null;
-		}
-		return convert( internalStack.peekFirst() );
+		return elements[head];
 	}
 
 	@Override
 	public T getRoot() {
-		if ( internalStack == null ) {
-			return null;
-		}
-		return convert( internalStack.peekLast() );
+		return elements[dec( tail )];
 	}
 
 	@Override
 	public int depth() {
-		if ( internalStack == null ) {
-			return 0;
+		int length = tail - head;
+		if ( length < 0 ) {
+			length += elements.length;
 		}
-		return internalStack.size();
+		return length;
 	}
 
 	@Override
 	public boolean isEmpty() {
-		if ( internalStack == null ) {
-			return true;
-		}
-		return internalStack.isEmpty();
+		return head == tail;
 	}
 
 	@Override
 	public void clear() {
-		if ( internalStack != null ) {
-			internalStack.clear();
+		circularClear( elements, head, tail );
+		head = tail = 0;
+	}
+
+	private static void circularClear(Object[] es, int i, int end) {
+		for ( int to = ( i <= end ) ? end : es.length; ; i = 0, to = end ) {
```

Comment:
I realize now that I gave you a bad suggestion, sorry.
In our particular case we don't really need a circular array; I had forgotten that `ArrayDeque` is based on a circular array, I guess because it allows adding elements either at the end or at the beginning of the queue (possibly needing to shift all entries).

but our use case is much simpler: we only have pop/push modification operations; this implies you should be able to simplify this all a little bit, as you don't need to keep track of both the head and the tail, only needing to ensure the array grows if there is not enough space: we always only insert new elements at the end.

The good news: you can indeed make this much more efficient than the existing implementation :)

Reply:
You're right, I was too focused on reproducing the previous behavior that I didn't stop and think if there was a better way of achieving this. The implementation will also be a lot easier tracking just the top of the stack. Thanks for all the suggestions!